Добавление стартера для логирования времени выполнения методов и интеграция в новый модуль

Colgens opened the pull request:
- Добавлен стартер `execution-time-logging-starter` для логирования времени выполнения методов.
- Добавлен модуль `FintechJuniorSpring` и произведена интеграция стартера.

Colgens committed:
feat: Создание модуля execution-time-logging-starter, добавление аннотации и аспекта

- Создан новый модуль execution-time-logging-starter.
- Добавлена аннотация LogExecutionTime для логирования времени выполнения методов.
- Добавлен аспект LogExecutionTimeAspect для обработки аннотации.
- Добавлен класс автоконфигурации ExecutionTimeLoggingAutoConfiguration.
- Добавлен файл spring.factories для автоматической конфигурации стартера.

Colgens committed:
feat: Создание модуля FintechJuniorSpring, интеграция стартера, добавление моделей, сервисов, репозиториев и других компонентов

- Добавлена зависимость на стартер execution-time-logging-starter в build.gradle.kts.
- Подключен стартер для логирования времени исполнения методов в основном проекте.
- Добавлены модели Category и Location.
- Добавлен интерфейс GenericRepository и его реализации в виде репозиториев CategoryRepository и LocationRepository.
- Добавлены сервисы CategoryService и LocationService.
- Добавлен сервис DataInitializationService для инициализации данных.
- Добавлены контроллеры CategoryController и LocationController.
- Добавлены классы конфигурации DataInitializationConfig, RestTemplateConfig.

kraskden commented on FintechJuniorSpring/src/main/java/com/tbank/fintechjuniorspring/kudago/repository/CategoryRepository.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    public Category save(Category category) {
+        if (category.getId() == null) {
```
Несколько проблем

1) Если я передам уже существующий ID, я перезатру уже существующую запись. Метод же должен делать только создание новой
2) Если я вызову save(id=2), save(), save(), то третий вызов save перезатрет первую запись, т.к. она создавалась с id = 2, а idGenerator на третьем вызове вернёт тоже 2. 

При POST не должно быть никакой перезаписи данных, только создание новых

kraskden replied:
Сам универсальный save можно оставить в репозитории, но на уровне сервиса надо решить эти проблемы

Colgens replied:
> Несколько проблем
> 
> 1. Если я передам уже существующий ID, я перезатру уже существующую запись. Метод же должен делать только создание новой
> 2. Если я вызову save(id=2), save(), save(), то третий вызов save перезатрет первую запись, т.к. она создавалась с id = 2, а idGenerator на третьем вызове вернёт тоже 2.
> 
> При POST не должно быть никакой перезаписи данных, только создание новых

Исправлено

kraskden commented on FintechJuniorSpring/src/main/java/com/tbank/fintechjuniorspring/kudago/service/CategoryService.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public Category getCategoryById(Long id) {
+        return categoryRepository.findById(id);
```
Если категории нету, то мы должны прокидывать статус 404
Делается это через проброс эксепшна с сервиса и отлавливание его в RestControllerAdvice

Аналогично обрабатываем кейс с обновлением или удалением несуществующей записи

Colgens replied:
> Если категории нету, то мы должны прокидывать статус 404 Делается это через проброс эксепшна с сервиса и отлавливание его в RestControllerAdvice
> 
> Аналогично обрабатываем кейс с обновлением или удалением несуществующей записи

Добавил исключения и хэндлер

Colgens committed:
fix: Исправление в build.gradle.kts для корректной работы со стартером, перенос конфигурации стартера, исправление логики репозиториев и сервисов для работы с уникальными id

feat: Добавлено кастомное исключение и хэндлер для корректного отображения статуса запросов

Colgens committed:
fix: Исправлен возвращаемый тип на ResponseEntity в хэндлере

kraskden commented on FintechJuniorSpring/src/main/java/com/tbank/fintechjuniorspring/kudago/service/CategoryService.java:
```diff
@@ -0,0 +1,65 @@
+    public Category createCategory(Category category) {
+        if (category.getId() == null) {
+            long id = idGenerator.getAndIncrement();
+            while (categoryRepository.findById(id) != null) {
```
Зачту, но в обычных рест апи при сохранении нельзя выбрать айдишник, он всегда автоматически берётся из генератора

Colgens replied:
Хорошо, я учту на будущее
